configure.ac

```diff
@@ -124,6 +124,49 @@ AC_COMPILE_IFELSE(
 		 AC_DEFINE([HAVE_FSCONFIG_CMD_CREATE], 1, [Define if FSCONFIG_CMD_CREATE is available])],
 		[AC_MSG_RESULT(no)])
 
+AC_MSG_CHECKING([for open_tree])
+AC_RUN_IFELSE(
+	[AC_LANG_SOURCE([[
+			#include <unistd.h>
+			#include <sys/syscall.h>
+			#include <errno.h>
+
+			int main(int argc, char **argv) {
+				int ret = syscall(__NR_open_tree);
+				return (ret >= 0 || errno != ENOSYS) ? 0 : 1;
+			}
+		]])],
+		[AC_MSG_RESULT(yes)
+		 AC_DEFINE([HAVE_OPEN_TREE], 1, [Define if OPEN_TREE is available])],
+		[AC_MSG_RESULT(no)])
+
+AC_MSG_CHECKING([for move_mount])
+AC_RUN_IFELSE(
+	[AC_LANG_SOURCE([[
+			#include <unistd.h>
+			#include <sys/syscall.h>
+			#include <errno.h>
+
+			int main(int argc, char **argv) {
+				int ret = syscall(__NR_move_mount);
+				return (ret >= 0 || errno != ENOSYS) ? 0 : 1;
+			}
+		]])],
+		[AC_MSG_RESULT(yes)
+		 AC_DEFINE([HAVE_MOVE_MOUNT], 1, [Define if MOVE_MOUNT is available])],
+		[AC_MSG_RESULT(no)])
+
+AC_MSG_CHECKING([for AT_RECURSIVE])
+AC_COMPILE_IFELSE(
+	[AC_LANG_SOURCE([[
+			#define _GNU_SOURCE
+			#include <fcntl.h>
+			int program = AT_RECURSIVE;
+		]])],
+		[AC_MSG_RESULT(yes)
+		 AC_DEFINE([HAVE_AT_RECURSIVE], 1, [Define if AT_RECURSIVE is available])],
+		[AC_MSG_RESULT(no)])
+
```

Comment:
could we replace the autoconf check with something similar to syscall_fsopen in linux.c?

In this way we will have to worry about the syscall being present only in one place

Reply:
Yup - working with autoconf definitely wasn't my favourite either - as you say in your other comment, I did this so I could put the feature info in the build flags so the test could be skipped on old kernels.

I did originally do this another way - I had `init.c` have another command "check-feature open_tree", which would try and call the `open_tree` syscall and either exit success or failure. Then, the python test harness does a `subprocess.call("./tests/init")` and checks the result.

Sounds like that might be better than these autoconf macros?

Reply:
If I do that and get rid of the autoconf macros, then I can have the `syscall_open_tree` helper return `ENOSYS` in the appropriate case like the others and handle that too as you suggest later on.

Reply:
(ah - one more reason for the autoconf - `AT_RECURSIVE` isn't defined in whatever build environment the LGTM tests run in. So I still need to ifdef around that. I _could_ do `#ifdef AT_RECURSIVE`, but there's no requirement for it to be a macro and not, say, an integer constant. Or maybe there's some buttons you can click in LGTM and make it more up to date)

Reply:
(ah - last observation before I go to bed - perhaps we could sidestep this by just not using `AT_RECURSIVE` at all. I added it because the bind mount currently does `MS_REC`, but I can't really think of a good reason to do that since there shouldn't be any further mounts underneath the notify  socket dir - just the socket file itself and nothing else)

Reply:
I am fine even if you define it if missing:

```
#ifndef AT_RECURSIVE
#  define AT_RECURSIVE		0x8000
#endif
```
and not have to worry if it is defined or not.

This should solve the issue with LGTM and in general with older build environments

Reply:
> I did originally do this another way - I had `init.c` have another command "check-feature open_tree", which would try and call the `open_tree` syscall and either exit success or failure. Then, the python test harness does a `subprocess.call("./tests/init")` and checks the result.

this sounds like a great idea